lib/rails_admin_import/importer.rb

```diff
@@ -0,0 +1,224 @@
+        return
+      end
+
+      perform_model_callback(object, :before_import_save, record)
+
+      if object.save
+        report_success(object, action)
+        perform_model_callback(object, :after_import_save, record)
+      else
+        report_error(object, action, object.errors.full_messages.join(", "))
+      end
+    end
+
+    def update_lookup
+      @update_lookup ||= if params[:update_if_exists] == "1"
+                           params[:update_lookup].to_sym
+                         end
+    end
+
+    attr_reader :results
+
+    def logger
+      @logger ||= ImportLogger.new
+    end
+
+    def report_success(object, action)
+      object_label = import_model.label_for_model(object)
+      message = I18n.t("admin.import.import_success.#{action}",
+                       :name => object_label)
+      logger.info "#{Time.now}: #{message}"
+      results[:success] << message
+    end
+
+    def report_error(object, action, error)
+      object_label = import_model.label_for_model(object)
+      message = I18n.t("admin.import.import_error.#{action}",
+                       :name => object_label,
+                       :error => error)
+      logger.info "#{Time.now}: #{message}"
+      results[:error] << message
+    end
+
+    def report_general_error(error)
+      message = I18n.t("admin.import.import_error.general", :error => error)
+      logger.info "#{Time.now}: #{message}"
+      results[:error] << message
+    end
+
+    def format_results
+      imported = results[:success]
+      not_imported = results[:error]
+      unless imported.empty?
+        results[:success_message] = format_result_message("successful", imported)
+      end
+      unless not_imported.empty?
+        results[:error_message] = format_result_message("error", not_imported)
+      end
+
+      results
+    end
+
+    def format_result_message(type, array)
+      result_count = "#{array.size} #{import_model.display_name.pluralize(array.size)}"
+      I18n.t("admin.flash.#{type}",
+             name: result_count,
+               action: I18n.t("admin.actions.import.done"))
+    end
+    
+    def perform_model_callback(object, method_name, record)
+      if object.respond_to?(method_name)
+        # Compatibility: Old import hook took 2 arguments.
+        # Warn and call with a blank hash as 2nd argument.
+        if object.method(method_name).arity == 2
+          report_old_import_hook(method_name)
+          object.send(method_name, record, {})
+        else
+          object.send(method_name, record)
+        end
+      end
+    end
+
+    def report_old_import_hook(method_name)
+      unless @old_import_hook_reported
+        error = I18n.t("admin.import.import_error.old_import_hook",
+                       model: import_model.display_name,
+                       method: method_name)
+        report_general_error(error)
+        @old_import_hook_reported = true
+      end
+    end
+
+    def find_or_create_object(record, update)
```

Comment:
Assignment Branch Condition size for find_or_create_object is too high. [17.97/15]